File: packages/imanikurd/src/dhikr.ts
import { loadData, loadDataAsync } from "./loadData.js";
import type { DhikrCategory, DhikrData, DhikrItem } from "./types.js";

let dhikrCache: DhikrData | null = null;

function getDhikrData(): DhikrData {
  if (!dhikrCache) {
    dhikrCache = loadData<DhikrData>("dhikr.json");
  }
  return dhikrCache;
}

async function getDhikrDataAsync(): Promise<DhikrData> {
  if (!dhikrCache) {
    dhikrCache = await loadDataAsync<DhikrData>("dhikr.json");
  }
  return dhikrCache;
}

/** Get all dhikr categories (morning, evening, after prayer, etc.) */
export function getDhikrCategories(): DhikrCategory[] {
  return getDhikrData().categories;
}

export async function getDhikrCategoriesAsync(): Promise<DhikrCategory[]> {
  return (await getDhikrDataAsync()).categories;
}

/** Get all dhikr items */
export function getDhikrItems(): DhikrItem[] {
  return getDhikrData().items;
}

export async function getDhikrItemsAsync(): Promise<DhikrItem[]> {
  return (await getDhikrDataAsync()).items;
}

/** Get dhikr items for a specific category */
export function getDhikrByCategory(categoryId: number): DhikrItem[] {
  return getDhikrData().items.filter((item) => item.categoryId === categoryId);
}

export async function getDhikrByCategoryAsync(categoryId: number): Promise<DhikrItem[]> {
  return (await getDhikrDataAsync()).items.filter((item) => item.categoryId === categoryId);
}

/** Get a single dhikr item by id */
export function getDhikrById(id: number): DhikrItem | undefined {
  return getDhikrData().items.find((item) => item.id === id);
}

export async function getDhikrByIdAsync(id: number): Promise<DhikrItem | undefined> {
  return (await getDhikrDataAsync()).items.find((item) => item.id === id);
}

/** Search dhikr by Kurdish or Arabic text */
export function searchDhikr(query: string): DhikrItem[] {
  const normalized = query.trim().toLowerCase();
  if (!normalized) return [];

  return getDhikrData().items.filter(
    (item) =>
      item.kurdish.toLowerCase().includes(normalized) ||
      item.arabic.includes(query.trim())
  );
}

export async function searchDhikrAsync(query: string): Promise<DhikrItem[]> {
  const normalized = query.trim().toLowerCase();
  if (!normalized) return [];

  return (await getDhikrDataAsync()).items.filter(
    (item) =>
      item.kurdish.toLowerCase().includes(normalized) ||
      item.arabic.includes(query.trim())
  );
}

/** Get category with its items */
export function getDhikrCategoryWithItems(categoryId: number) {
  const category = getDhikrData().categories.find((c) => c.id === categoryId);
  if (!category) return null;

  return {
    category,
    items: getDhikrByCategory(categoryId),
  };
}

export async function getDhikrCategoryWithItemsAsync(categoryId: number) {
  const data = await getDhikrDataAsync();
  const category = data.categories.find((c) => c.id === categoryId);
  if (!category) return null;

  return {
    category,
    items: data.items.filter((item) => item.categoryId === categoryId),
  };
}

export type { DhikrCategory, DhikrData, DhikrItem };

File: packages/imanikurd_flutter/lib/src/dhikr.dart
import 'load_data.dart';
import 'models.dart';

DhikrData? _dhikrCache;

Future<DhikrData> _getDhikrData() async {
  _dhikrCache ??= DhikrData.fromJson(
    await loadDataAsync<Map<String, dynamic>>('dhikr.json'),
  );
  return _dhikrCache!;
}

Future<List<DhikrCategory>> getDhikrCategories() async => (await _getDhikrData()).categories;

Future<List<DhikrItem>> getDhikrItems() async => (await _getDhikrData()).items;

Future<List<DhikrItem>> getDhikrByCategory(int categoryId) async {
  final data = await _getDhikrData();
  return data.items.where((item) => item.categoryId == categoryId).toList();
}

Future<DhikrItem?> getDhikrById(int id) async {
  final data = await _getDhikrData();
  for (final item in data.items) {
    if (item.id == id) return item;
  }
  return null;
}

Future<List<DhikrItem>> searchDhikr(String query) async {
  final normalized = query.trim().toLowerCase();
  if (normalized.isEmpty) return [];
  final data = await _getDhikrData();
  return data.items
      .where(
        (item) =>
            item.kurdish.toLowerCase().contains(normalized) ||
            item.arabic.contains(query.trim()),
      )
      .toList();
}

Future<DhikrCategoryWithItems?> getDhikrCategoryWithItems(int categoryId) async {
  final data = await _getDhikrData();
  DhikrCategory? category;
  for (final c in data.categories) {
    if (c.id == categoryId) {
      category = c;
      break;
    }
  }
  if (category == null) return null;
  return DhikrCategoryWithItems(
    category: category,
    items: data.items.where((item) => item.categoryId == categoryId).toList(),
  );
}
